### ISSUE TITLE
SSH key management changes ownership permissions of authorized_key file

### ISSUE THREAD

**Issue by ColonelPanicks** [alces-software/overware#1]
I had created an SSH keyfile with a single key in with the following permissions:
```
[root@gateway1 [gateway1] ~]# ls -ld ~login/.ssh/authorized_keys
-rw-r--r--. 1 login login 1179 Jan 21 14:52 /appliance/login/.ssh/authorized_keys
```

Overware is setup to use this file for SSH key management. If I delete a key from the file it does get removed but the keyfile permissions change:
```
[root@gateway1 [gateway1] ~]# ls -ld ~login/.ssh/authorized_keys
-rw-------. 1 root root 0 Jan 21 14:54 /appliance/login/.ssh/authorized_keys
```

This will break key authorisation for the user. Instead, the permissions of the keyfile should be preserved.

**Comment by ColonelPanicks** [comment]
## Further investigation

I have a keyfile setup with the following permissions and content:
```
[root@gateway1 [gateway1] ~]# cat /appliance/login/.ssh/authorized_keys
ssh-rsa mykey stu@test
ssh-rsa anotherkey stu@machine
[root@gateway1 [gateway1] ~]# ls -ld /appliance/login/.ssh/authorized_keys
-rw-r--r-- 1 login login 54 Jan 22 19:02 /appliance/login/.ssh/authorized_keys
[root@gateway1 [gateway1] ~]# ls -ld /appliance/login/.ssh
drwx------ 2 login login 4096 Jan 22 19:02 /appliance/login/.ssh
```

I can see the keys from the app:
<img width="1432" alt="screenshot 2019-01-22 at 19 04 15" src="https://user-images.githubusercontent.com/27725639/51559002-b7cbb400-1e78-11e9-94a2-0517384ca385.png">

But if I add a key I get the success message and the key doesn't appear in the app or in the authorized_keys file.

However, if I delete a key, it gets removed from the file and then screws the permissions
```
[root@gateway1 [gateway1] ~]# cat /appliance/login/.ssh/authorized_keys
ssh-rsa anotherkey stu@machine
[root@gateway1 [gateway1] ~]# ls -ld /appliance/login/.ssh/authorized_keys
-rw------- 1 root root 31 Jan 22 19:06 /appliance/login/.ssh/authorized_keys
```

Even after removing a key and seeing the permissions change I am unable to add new keys. All that I get is the success message and cannot see the key in either the app or the file.

**Comment by ColonelPanicks** [comment]
Below is the function for creating a new key entry
https://github.com/alces-software/overware/blob/master/app/controllers/keys_controller.rb#L9-L21

From the looks of things this **only** updates the temp file (`tmp/keys`) and doesn't actually do anything to the configured SSH key file in the environment.

This becomes very apparent when compared with the usage of a tempfile by the delete function
https://github.com/alces-software/overware/blob/master/app/controllers/keys_controller.rb#L23-L36

This also explains why the `create` always gives a success message even though the environment SSH file isn't touched.

In regards to the permissions issues - [it looks like `FileUtils.mv` supports preserving permissions with `:preserve => true`](https://ruby-doc.org/stdlib-2.4.1/libdoc/fileutils/rdoc/FileUtils.html#module-FileUtils-label-Module+Functions)

I have pushed branch `fix/file-updates-and-permissions` with my WIP on a solution.

**Comment by atoghill** [comment]
Resolved as of #2.
